src/allocator_defrag.c

```diff
@@ -0,0 +1,379 @@
+    jeBinInfo *bin_info;           /* Array of `jeBinInfo` structs, one for each bin. */
+    jeMallctlKey util_batch_query; /* Key to query batch utilization information. */
+    jeMallctlKey epoch;            /* Key to trigger statistics sync between threads. */
+} jeBinsConf;
+
+/* Represents the latest usage statistics for a jemalloc bin.
+ *
+ * This struct tracks the current usage of a bin, including the number of slabs
+ * and regions, and calculates the number of full slabs from other fields.
+ */
+typedef struct jeBusage {
+    unsigned long curr_slabs;         /* Current number of slabs in the bin. */
+    unsigned long curr_nonfull_slabs; /* Current number of non-full slabs in the bin. */
+    unsigned long curr_full_slabs;    /* Current number of full slabs in the bin (calculated from other fields). */
+    unsigned long curr_regs;          /* Current number of regions in the bin. */
+} jeBusage;
+
+
+static int defrag_supported = 0;
+static size_t jemalloc_quantum = 0;
+static jeBinsConf arena_bin_conf = {0, NULL, {{0}, 0}, {{0}, 0}};
+static jeBusage *curr_usage = NULL;
+
+
+/* -----------------------------------------------------------------------------
+ * Alloc/Free API that are cooperative with defrag
+ * -------------------------------------------------------------------------- */
+
+/* Allocation and free functions that bypass the thread cache
+ * and go straight to the allocator arena bins.
+ * Currently implemented only for jemalloc. Used for online defragmentation.
+ */
+void *allocatorDefragAlloc(size_t size) {
+    void *ptr = je_mallocx(size, MALLOCX_TCACHE_NONE);
+    return ptr;
+}
+void allocatorDefragFree(void *ptr, size_t size) {
+    if (ptr == NULL) return;
+    je_sdallocx(ptr, size, MALLOCX_TCACHE_NONE);
+}
+
+/* -----------------------------------------------------------------------------
+ * Helper functions for jemalloc translation between size and index
+ * -------------------------------------------------------------------------- */
+#define LG_QUANTOM_8_FIRST_POW2 3
+#define SIZE_CLASS_GROUP_SZ 4
+
+#define LG_QUANTOM_OFFSET_3 ((64 >> LG_QUANTOM_8_FIRST_POW2) - 1)
+
+#define getBinindNormal(_sz, _offset, _last_sz_pow2)                                                             \
+    ((SIZE_CLASS_GROUP_SZ - (((1 << (_last_sz_pow2)) - (_sz)) >> ((_last_sz_pow2) - LG_QUANTOM_8_FIRST_POW2))) + \
+     (((_last_sz_pow2) - (LG_QUANTOM_8_FIRST_POW2 + 3)) - 1) * SIZE_CLASS_GROUP_SZ + (_offset))
+
+/* Get the bin index in bin array from the reg_size.
+ *
+ * these are reverse engineered mapping of reg_size -> binind. We need this information because the utilization query
+ * returns the size of the buffer and not the bin index, and we need the bin index to access it's usage information
+ *
+ * Note: In case future PR will return the binind (that is better API anyway) we can get rid of
+ * these conversion functions
+ */
+inline unsigned jeSize2BinIndexLgQ3(size_t sz) {
```

Comment:
```suggestion
static inline unsigned jeSize2BinIndexLgQ3(size_t sz) {
```

Whenever we use `inline`, use `static inline` instead, or only `static`. Only `inline` by itself is a compiler hint that is more or less ignored and doesn't have a clear meaning.